packages/api/src/external/carequality/document/query-documents.ts

```diff
@@ -0,0 +1,213 @@
+      cqLinks: cqPatientData?.data.links ?? [],
+    });
+
+    await iheGateway.startDocumentsQuery(docQueryRequest);
+
+    // TODO: CHANGE THIS TO THE STRATEGY RAMIL USED: RACE CONTROL
+    const results = await pTimeout(
+      getDocumentQueryResult({ requestId }),
+      DOCUMENT_QUERY_TIMEOUT.asMilliseconds(),
+      "handleDocQueryResponse function timed out!"
+    );
+
+    await handleDocQueryResults(requestId, patient.id, patient.cxId, results, log);
+  } catch (error) {
+    const msg = `Failed to query and process documents in Carequality.`;
+    console.log(`${msg}. Error: ${errorToString(error)}`);
+
+    const updatedPatient = await appendDocQueryProgress({
+      patient: { id: patient.id, cxId: patient.cxId },
+      downloadProgress: { status: "failed" },
+      requestId,
+      source: MedicalDataSource.COMMONWELL,
+    });
+
+    const downloadProgressHasFailed =
+      updatedPatient.data.documentQueryProgress?.download?.status === "failed";
+
+    if (downloadProgressHasFailed) {
+      processPatientDocumentRequest(
+        patient.cxId,
+        patient.id,
+        "medical.document-download",
+        MAPIWebhookStatus.failed
+      );
+    }
+
+    capture.message(msg, {
+      extra: {
+        context: `cq.queryAndProcessDocuments`,
+        error,
+        patientId: patient.id,
+        requestId,
+      },
+      level: "error",
+    });
+    throw error;
+  }
+}
+
+export async function handleDocQueryResults(
+  requestId: string,
+  patientId: string,
+  cxId: string,
+  docQueryResults: DocumentQueryResult[],
+  log = console.log
+): Promise<void> {
+  const docRefs = combineDocRefs(docQueryResults);
+
+  const docsToDownload = await getNonExistentDocRefs(docRefs, patientId, cxId);
+
+  const convertibleDocCount = docsToDownload.filter(doc => isConvertible(doc.contentType)).length;
+
+  log(`I have ${docsToDownload.length} docs to download (${convertibleDocCount} convertible)`);
+
+  appendDocQueryProgress({
+    patient: { id: patientId, cxId: cxId },
+    downloadProgress: {
+      status: "processing",
+      total: docsToDownload.length,
+    },
+    convertProgress: {
+      status: "processing",
+      total: convertibleDocCount,
+    },
+    requestId,
+    source: MedicalDataSource.COMMONWELL,
+  });
+
+  // TODO - INTRODUCED WHEN IMPLEMENTING CQ DOC RETRIEVAL
+  // downloadDocs(docsToDownload, patientId, cxId, requestId);
+}
+
+function combineDocRefs(docQueryResults: DocumentQueryResult[]): DocumentReference[] {
+  return docQueryResults.reduce((acc: DocumentReference[], curr) => {
+    return [...acc, ...curr.data.documentReference];
+  }, []);
+}
+
+const getNonExistentDocRefs = async (
+  documents: DocumentReference[],
+  patientId: string,
+  cxId: string
+): Promise<DocumentReference[]> => {
+  const [{ existingDocRefs, nonExistingDocRefs }, FHIRDocRefs] = await Promise.all([
+    filterOutExistingDocRefsS3(documents, patientId, cxId),
+    getDocRefsFromFHIR(cxId, patientId),
+  ]);
```

Comment:
This, and each function, is pretty similar to what we have for CW. Can't we rewrite this and have both using the same logic?

Reply:
ditto